plugins/in_systemd/systemd.c

```diff
@@ -211,14 +214,29 @@ static int in_systemd_collect(struct flb_input_instance *ins,
                 key++;
                 length--;
             }
+
             sep = strchr(key, '=');
             if (sep == NULL) {
                 skip_entries++;
                 continue;
             }
+
             len = (sep - key);
             msgpack_pack_str(&mp_pck, len);
-            msgpack_pack_str_body(&mp_pck, key, len);
+
+            if (ctx->lowercase == FLB_TRUE) {
+                tmp = flb_sds_create_len(key, len);
```

Comment:
I was wondering if we could re-use the original systemd buffer (pointed by `key`) to perform the lowercase conversion, can you check it ?

if the previous step is not possible, we might use a pre-allocated buffer for key conversion to avoid a tons of malloc/free on each record iteration. I see in the docs that usually records gets truncated at 64kb:

- https://manpages.debian.org/testing/libsystemd-dev/sd_journal_enumerate_data.3.en.html#DESCRIPTION

Reply:
Thank you for taking the time to review this pull request!

I've seen the manpage saying that ”returned data is in a read-only memory map” so chose not to overwritten the data directly; however, I see your point, so how about pre-allocating a buffer whose size is equal to the threshold from `sd_journal_get_data_threshold()` and increase it when it’s insufficient to hold the entire key? I’ve pushed the branch trying to reflect this.

> This threshold is a hint only: it indicates that the client program is interested only in the initial parts of the data fields, up to the threshold in size — but the library might still return larger data objects. That means applications should not rely exclusively on this setting to limit the size of the data fields returned, but need to apply an explicit size limit on the returned data as well.

Reply:
that looks pretty good :) , thank you!